From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Martijn Muijsers <martijnmuijsers@live.nl>
Date: Sun, 30 Jul 2023 18:08:52 +0200
Subject: [PATCH] Client perspective - Deep replacements - Method to send
 updates

Adds a method to send the necessary updates to a player when
the way they observe deep replaceables may have changed.

License: GPL-3.0 (https://www.gnu.org/licenses/gpl-3.0.html)
Fiddle - https://fiddlemc.org

diff --git a/src/main/java/net/minecraft/server/level/ServerPlayer.java b/src/main/java/net/minecraft/server/level/ServerPlayer.java
index 6b3520b0a2b5a03e5be440b6bea96070fe0cb7a5..3efbe919ff25fd738ed6ba01f297f0f8bc49a13a 100644
--- a/src/main/java/net/minecraft/server/level/ServerPlayer.java
+++ b/src/main/java/net/minecraft/server/level/ServerPlayer.java
@@ -2783,4 +2783,32 @@ public class ServerPlayer extends Player {
     }
     // Fiddle end - client perspective - player
 
+    // Fiddle start - client perspective - deep replacements - send updates
+    /**
+     * Sends all necessary packets to this player to update any deep replaceable content
+     * after the content that the client should observe may have changed.
+     * <br>
+     * This method should be called whenever something about the player or their client changes that influences
+     * deep replacements.
+     */
+    public void sendUpdatesForDeepReplaceableContent(boolean itemsMayHaveChanged, boolean translationsMayHaveChanged) {
+
+        // Items contain items, so we make sure that this implication is set correctly
+        itemsMayHaveChanged |= translationsMayHaveChanged;
+
+        if (itemsMayHaveChanged) {
+            // Resend the player's own inventory contents
+            this.getBukkitEntity().updateInventory();
+        }
+
+        // TODO if itemsMayHaveChanged, re-send tracked entity synched data
+        // TODO if itemsMayHaveChanged, re-send all visible block entity data
+        // TODO if itemsMayHaveChanged, re-send contents of other inventory that player has open
+        // TODO if translationsMayHaveChanged, update title of inventory that player has open
+        // TODO if translationsMayHaveChanged, update other visible texts to player like boss bar text
+        // TODO if translationsMayHaveChanged, update book content if they have one open maybe?
+
+    }
+    // Fiddle end - client perspective - deep replacements - send updates
+
 }
